File: Backend/Application/Common/PersonNameNormalizer.cs
using System.Text.RegularExpressions;

namespace Application.Common;

public static class PersonNameNormalizer
{
    private static readonly HashSet<string> LowercaseParticles = new(StringComparer.Ordinal)
    {
        "da",
        "de",
        "do",
        "das",
        "dos",
        "e",
        "di",
        "du",
        "van",
        "von"
    };

    public static string Normalize(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return string.Empty;
        }

        var words = Regex.Replace(value.Trim(), @"\s+", " ")
            .ToLowerInvariant()
            .Split(' ', StringSplitOptions.RemoveEmptyEntries);

        return string.Join(' ', words.Select((word, index) =>
            index > 0 && LowercaseParticles.Contains(word)
                ? word
                : NormalizeWord(word)));
    }

    private static string NormalizeWord(string word)
    {
        return string.Join('\'', word.Split('\'').Select(NormalizePart));
    }

    private static string NormalizePart(string part)
    {
        if (string.IsNullOrEmpty(part))
        {
            return part;
        }

        if (part.StartsWith("mc", StringComparison.Ordinal) && part.Length > 2)
        {
            return $"Mc{CapitalizeFirst(part[2..])}";
        }

        return CapitalizeFirst(part);
    }

    private static string CapitalizeFirst(string value)
    {
        return string.IsNullOrEmpty(value)
            ? value
            : $"{char.ToUpperInvariant(value[0])}{value[1..]}";
    }
}

File: Frontend/Inventario/src/app/shared/person-name-normalizer.ts
const LOWERCASE_PARTICLES = new Set(['da', 'de', 'do', 'das', 'dos', 'e', 'di', 'du', 'van', 'von']);

export function normalizePersonName(value: string): string {
  const words = value
    .trim()
    .replace(/\s+/g, ' ')
    .toLowerCase()
    .split(' ')
    .filter(Boolean);

  return words
    .map((word, index) => {
      if (index > 0 && LOWERCASE_PARTICLES.has(word)) {
        return word;
      }

      return normalizeNameWord(word);
    })
    .join(' ');
}

export function normalizePersonNameInput(value: string): string {
  const hasTrailingSpace = /\s$/.test(value);
  const normalized = normalizePersonName(value);

  return hasTrailingSpace && normalized ? `${normalized} ` : normalized;
}

function normalizeNameWord(word: string): string {
  return word
    .split("'")
    .map((part) => normalizeNamePart(part))
    .join("'");
}

function normalizeNamePart(part: string): string {
  if (!part) {
    return part;
  }

  if (part.startsWith('mc') && part.length > 2) {
    return `Mc${capitalizeFirst(part.slice(2))}`;
  }

  return capitalizeFirst(part);
}

function capitalizeFirst(value: string): string {
  return `${value.charAt(0).toUpperCase()}${value.slice(1)}`;
}
